### PR TITLE
Implement tasks 1-6

### PR DESCRIPTION
**Pull request by pzanevskiy**

### CODE REVIEW

File: 01.multithreading/MultiThreading.Task2.Chaining/Program.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            var fourthTask = thirdTask.ContinueWith(prevTask =>
+            {
+                var array = prevTask.Result;
+                var average = array.Average();
+
+                Console.WriteLine("Fourth task: {0}", average);
+            });
+
+            fourthTask.Wait();
+
```
Comment (vadzim-m) [inline 01.multithreading/MultiThreading.Task2.Chaining/Program.cs:75]: It is a little bit difficult to read the whole main method, lots of detail are on the same level of abstraction. Going to discuss it on the next session.  Nothing to change.

File: 01.multithreading/MultiThreading.Task2.Chaining/Program.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            });
+
+            var fourthTask = thirdTask.ContinueWith(prevTask =>
+            {
+                var array = prevTask.Result;
+                var average = array.Average();
+
+                Console.WriteLine("Fourth task: {0}", average);
+            });
+
```
Comment (vadzim-m) [inline 01.multithreading/MultiThreading.Task2.Chaining/Program.cs:71]: Unused Task variables were detected)

File: 01.multithreading/MultiThreading.Task2.Chaining/Program.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                    array[i] = random.Next(0, 100);
+                }
+
+                Console.WriteLine("First task: {0}", string.Join(", ", array));
+                return array;
+            });
+
+            var secondTask = firstTask.ContinueWith(prevTask =>
+            {
+                var random = new Random();
```
Comment (vadzim-m) [inline 01.multithreading/MultiThreading.Task2.Chaining/Program.cs]: Random instance can be shared with two methods.

File: 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+            ProcThread(StateNumber);
+            ProcThreadPool(StateNumber);
+        }
+
+        private static void ProcThread(object obj)
+        {
+            var number = (int)obj;
+            Console.WriteLine("Threads - " + number);
+            if (number > 0)
```
Comment (AliceInHell) [inline 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs:41]: Take a look how many thread did you create) Seems 11, the right condition is number > 1.

File: 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            Console.WriteLine("Each thread should be with the same body and receive a state with integer number, decrement it, print and pass as a state into the newly created thread.");
+            Console.WriteLine("Implement all of the following options:");
+            Console.WriteLine();
+            Console.WriteLine("- a) Use Thread class for this task and Join for waiting threads.");
+            Console.WriteLine("- b) ThreadPool class for this task and _semaphore for waiting threads.");
+
+            Console.WriteLine();
+
+            ProcThread(StateNumber);
+            ProcThreadPool(StateNumber);
```
Comment (vadzim-m) [inline 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs:36]: Your main thread can not wait for executing children using semaphore. Look at the console output. ThreadPool - 10 and console app ends.

File: 01.multithreading/MultiThreading.Task5.Threads.SharedCollection/Program.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+        static void Main(string[] args)
+        {
+            Console.WriteLine("5. Write a program which creates two threads and a shared collection:");
+            Console.WriteLine("the first one should add 10 elements into the collection and the second should print all elements in the collection after each adding.");
+            Console.WriteLine("Use Thread, ThreadPool or Task classes for thread creation and any kind of synchronization constructions.");
+            Console.WriteLine();
+
+            var addingTask = AddingTask();
+            var printingTask = PrintingTask();
```
Comment (vadzim-m) [inline 01.multithreading/MultiThreading.Task5.Threads.SharedCollection/Program.cs:27]: Your code didn't output the last collection with 10 elements. They even are not synchronized and print task can output same collection twice.

Comment (vadzim-m) [reply 01.multithreading/MultiThreading.Task5.Threads.SharedCollection/Program.cs:27]: Please, look at the difference between Semaphore and SemaphoreSlim. (just replace your semaphore type and run again)

Comment (pzanevskiy) [reply 01.multithreading/MultiThreading.Task5.Threads.SharedCollection/Program.cs:27]: Fixed

File: 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs (hunk 5 above)
Comment (pzanevskiy) [reply 01.multithreading/MultiThreading.Task4.Threads.Join/Program.cs:36]: I'm just executing my code and everything was successful🙃

**Pull request merged**